src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -132,102 +154,165 @@ public virtual byte[] GetValue()
 
         internal static void TransformControls(DirectoryControl[] controls)
         {
-            for (int i = 0; i < controls.Length; i++)
+            Span<byte> attributeNameScratchSpace = stackalloc byte[AttributeNameStackAllocationThreshold];
+
+            try
             {
-                Debug.Assert(controls[i] != null);
-                byte[] value = controls[i].GetValue();
-                if (controls[i].Type == "1.2.840.113556.1.4.319")
+                for (int i = 0; i < controls.Length; i++)
                 {
-                    // The control is a PageControl.
-                    object[] result = BerConverter.Decode("{iO}", value);
-                    Debug.Assert((result != null) && (result.Length == 2));
+                    Debug.Assert(controls[i] != null);
+                    byte[] value = controls[i]._directoryControlValue ?? Array.Empty<byte>();
+                    Span<byte> asnSpan = value;
+                    bool asnReadSuccessful;
 
-                    int size = (int)result[0];
-                    // user expects cookie with length 0 as paged search is done.
-                    byte[] cookie = (byte[])result[1] ?? Array.Empty<byte>();
-
-                    PageResultResponseControl pageControl = new PageResultResponseControl(size, cookie, controls[i].IsCritical, controls[i].GetValue());
-                    controls[i] = pageControl;
-                }
-                else if (controls[i].Type == "1.2.840.113556.1.4.1504")
-                {
-                    // The control is an AsqControl.
-                    object[] o = BerConverter.Decode("{e}", value);
-                    Debug.Assert((o != null) && (o.Length == 1));
+                    if (controls[i].Type == "1.2.840.113556.1.4.319")
+                    {
+                        // The control is a PageControl, as described in RFC 2696.
```

Comment:
RFC 2696 does not describe a `PageControl` structure (that sequence of characters does not appear in the RFC).

This looks like it might be a `realSearchControlValue` from https://datatracker.ietf.org/doc/html/rfc2696#section-2.

Reply:
Yes, agreed. The comments are supposed to refer to the DirectoryControl-derivatives being constructed, but this isn't accurate across the board. I'll change them and include the name of the structure they're being parsed from for clarity.